### PR TITLE
Skipping scale-up if MachineTypeNotAvailableAnnotation is present on MachineDeployment

### PR DESCRIPTION
**Pull request by hardikdr**
**What this PR does / why we need it**:  This PR enables autoscaler to skip the scale-up if `MachineTypeNotAvailable` annotation is present on the machine-deployment. 
This annotation is set by the MCM, when certain machine-types are not out of capacity and can't be further scaled-up.

As an outcome of this change, autoscaler will pick up alternate node-group if one is not capable of scaling.  

Related to : https://github.com/gardener/autoscaler/issues/35
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Autoscaler avoids scaling-up the MachineDeployment if MachineTypeNotAvailableAnnotation is present
```

### CODE REVIEW

**Comment by prashanth26** [comment]
Hi @hardikdr ,

Were these changes tested with a real setup?

**Comment by hardikdr** [comment]
Yes, I have tested this with the related changes at MCM [#454](https://github.com/gardener/machine-controller-manager/pull/454), works fine.

**Review by prashanth26 — APPROVED** [review]
hold

File: cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go [hunk 1]
```diff
@@ -38,6 +39,9 @@ import (
 const (
 	// ProviderName is the cloud provider name for MCM
 	ProviderName = "mcm"
+
+	//MachineTypeNotAvailableAnnotation is an annotation put by MCM on machine-deployment if certain machine-types are not available.
+	MachineTypeNotAvailableAnnotation = "machine.sapcloud.io/machine-type-not-available"
```
Comment (prashanth26) [inline cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go:44]: The propagation of this label from machine/machineDeployment controller till the autoscaler seems like a hack. Is there a more elegent solution here? If not can we keep this hack only at the autoscaler for now. By making the autoscaler search for failed machine patterns in the failedMachines field?

Comment (hardikdr) [reply cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go:44]: Yes sure, I will give it a try with changes only at CA, and not touch MCM that definitely be the less complicated solution.

**Comment by jsravn** [comment]
Is this the correct approach? From looking at it, upstream autoscaler already handles this but you have to return unregistered instances in the autoscaler cloudprovider interface, which this fork isn't doing properly.

**Comment by hardikdr** [comment]
@jsravn That's true, and also one of the reasons why I stopped working on this approach for now.
The interface in this version of CA, isn't mature enough, and I am rebasing the autoscaler to the latest version here: https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider. I'll then be looking at this topic again and find a more natural way to fix it.

File: cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go (hunk 1 above)
Comment (hardikdr) [reply cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go:44]: We need to re-investigate the overall approach.

**Comment by jsravn** [comment]
@hardikdr Ah cool. I was also going to try rebasing. I should have a PR soon which fixes it to return unregistered instances - so at least this functionality will work as expected in the existing code (if it fails to scale up the machinedeployment, autoscaler will give up after max-node-provision-time and try the next one).

**Comment by hardikdr** [comment]
Sounds great. We are discussing this issue at length here: https://github.com/gardener/autoscaler/issues/35.

Also, you might want to make sure, your changes are on top of the `machine-controller-manager-provider`, which is also a default branch now.

**Comment by jsravn** [comment]
@hardikdr FYI I opened a PR at https://github.com/gardener/autoscaler/pull/58.

**Comment by prashanth26** [comment]
We no longer need this PR right? Thanks to this - https://github.com/gardener/autoscaler/pull/58?

**Comment by hardikdr** [comment]
No, we don't need this anymore. 
/close

**Pull request closed without merging**